stress test for tail calls (#2518)
microsoft/ebpf-for-windows: tests/stress/readme.md

- Extension restart enabled.
- Delay of 250 ms between successive extension restarts.

> ## 1.6. bindmonitor_tailcall_invoke_program_test

Reviewer: ```suggestion
## 1.6. bindmonitor_tail_call_invoke_program_test
```

> This test first loads a specific native eBPF program. It then loads all the 32 tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the port. The test set up ensures that the `thread_index` passed in each `thread_context` is unique to that thread.

Reviewer: ```suggestion
This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the port. The test set up ensures that the `thread_index` passed in each `thread_context` is unique to that thread.
```
avoid hard coding the number 32 in other places.  It might change in the future and we want to minimize the number of places to change in such a case since they're easy to miss.

> This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the port. The test set up ensures that the `thread_index` passed in each `thread_context` is unique to that thread.

Reviewer 2: nit: set up (verb), setup (noun)

Reviewer: +1 
```suggestion
This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the port. The test setup ensures that the `thread_index` passed in each `thread_context` is unique to that thread.
```

> This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the same port continuously in a loop. The test setup ensures that the `thread_index` passed in each `thread_context` is unique to that thread.

Author: Each thread has a unique thread index which starts at 0 and that can be passed in the thread context. These thread indexes can be used for debugging when the thread performs the tail call functionality.

> This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the same port continuously in a loop. The test setup ensures that the `thread_index` passed in each `thread_context` is unique to that thread.

Reviewer 3: Yes. I'm aware of that.  That verbiage needs to be in this _document_.  As-is, that sentence does not tie into the preceding content and is very confusing, especially for a first-time reader.

More importantly, apart from the debugging usage, a more critical usage of that index is to ensure that each thread gets a unique port (base_port + thread_index).  Please mention this in the document update.

> This test first loads a specific native eBPF program. It then loads all the MAX_TAIL_CALL_CNT tail call programs and updates the program array map. It then creates the specified # of threads where each thread attempts a TCP 'bind' to the same port continuously in a loop. The test setup ensures that the `thread_index` passed in each `thread_context` is unique to that thread.

Author: Added.

> This causes the invocation of the in-kernel eBPF tail call programs to be executed in sequence. The last tail call program returns PERMIT verdict.

Reviewer: ```suggestion
This causes the invocation of the in-kernel eBPF tail call programs to be executed in sequence. The last tail call program returns a PERMIT verdict.
```
grammar

This test can be run with or without the extension restart option.

Sample command line invocations:

> ### 1.6.1. `ebpf_stress_test_km bindmonitor_tailcall_invoke_program_test`
> - Uses default values for all supported options.
> 
> ### 1.6.2. `ebpf_stress_test_km -tt=32 -td=15 -vo=true -er=true -erd=250 bindmonitor_tailcall_invoke_program_test`

Reviewer: ```suggestion
### 1.6.1. `ebpf_stress_test_km bindmonitor_tail_call_invoke_program_test`
- Uses default values for all supported options.

### 1.6.2. `ebpf_stress_test_km -tt=32 -td=15 -vo=true -er=true -erd=250 bindmonitor_tail_call_invoke_program_test`
```
